python_modules/dagster-ext/dagster_ext/__init__.py

```diff
@@ -1,32 +1,732 @@
+
+    @property
+    def extras(self) -> Mapping[str, Any]:
+        return self._data["extras"]
+
+    # ##### WRITE
+
+    def report_asset_metadata(
+        self, label: str, value: Any, asset_key: Optional[str] = None
+    ) -> None:
+        asset_key = _resolve_optionally_passed_asset_key(
+            self._data, asset_key, "report_asset_metadata"
+        )
+        label = _assert_param_type(label, str, "report_asset_metadata", "label")
+        value = _assert_param_json_serializable(value, "report_asset_metadata", "value")
+        self._write_message(
+            "report_asset_metadata", {"asset_key": asset_key, "label": label, "value": value}
+        )
+
+    def report_asset_data_version(self, data_version: str, asset_key: Optional[str] = None) -> None:
+        asset_key = _resolve_optionally_passed_asset_key(
+            self._data, asset_key, "report_asset_data_version"
+        )
+        data_version = _assert_param_type(
+            data_version, str, "report_asset_data_version", "data_version"
+        )
+        self._write_message(
+            "report_asset_data_version", {"asset_key": asset_key, "data_version": data_version}
+        )
+
+    def log(self, message: str, level: str = "info") -> None:
+        message = _assert_param_type(message, str, "log", "asset_key")
+        level = _assert_param_value(level, ["info", "warning", "error"], "log", "level")
+        self._write_message("log", {"message": message, "level": level})
+
+
+# ########################
+# ##### PUBLIC API
+# ########################
+
+# Defining `__all__` would be redundant for purposes of formalizing the public API, since
+# type-checkers will count any non-underscored symbol defined in this module as public. However, it
+# is useful to maintain a whitelist of intended-to-be-public symbols for development and
+# documentation purposes.
+
+# ExtContext as ExtContext,
+# init_dagster_ext as init_dagster_ext,
+# ExtContextLoader as ExtContextLoader,
+# ExtBlobStoreMessageWriter as ExtBlobStoreMessageWriter,
+# ExtBlobStoreMessageWriterChannel as ExtBlobStoreMessageWriterChannel,
+# ExtMessageWriter as ExtMessageWriter,
+# ExtMessageWriterChannel as ExtMessageWriterChannel,
+# ExtParamLoader as ExtParamLoader,
+# ExtDefaultContextLoader as ExtDefaultContextLoader,
+# ExtDefaultMessageWriter as ExtDefaultMessageWriter,
+# ExtEnvContextLoader as ExtEnvContextLoader,
+# ExtS3MessageWriter as ExtS3MessageWriter,
+# DAGSTER_EXT_ENV_KEYS as DAGSTER_EXT_ENV_KEYS,
+# ExtContextData as ExtContextData,
+# ExtDataProvenance as ExtDataProvenance,
+# ExtExtras as ExtExtras,
+# ExtMessage as ExtMessage,
+# ExtParams as ExtParams,
+# ExtPartitionKeyRange as ExtPartitionKeyRange,
+# ExtTimeWindow as ExtTimeWindow,
+# DagsterExtError as DagsterExtError,
+# DagsterExtWarning as DagsterExtWarning,
+# decode_env_var as decode_env_var,
+# encode_env_var as encode_env_var,
+# is_dagster_orchestration_active as is_dagster_orchestration_active,
```

Comment:
hmmm very skeptical of the utility of this. almost certainly will fall out of date

Reply:
fair, it was more useful while I was picking out the methods to underscore than it is now, removed